### PR TITLE
Fix shard retry mechanism in `push_to_hub`

### PR DESCRIPTION
**Pull request by mariosasko**
When it fails, `preupload_lfs_files` throws a [`RuntimeError`](https://github.com/huggingface/huggingface_hub/blob/5eefebee2c150a2df950ab710db350e96c711433/src/huggingface_hub/_commit_api.py#L402) error and chains the original HTTP error. This PR modifies the retry mechanism's error handling to account for that.

Fix https://github.com/huggingface/datasets/issues/6392

### CODE REVIEW

**Comment by mariosasko** [comment]
@Wauplin Maybe `504` should be added to the `retry_on_status_codes` tuple [here](https://github.com/huggingface/huggingface_hub/blob/5eefebee2c150a2df950ab710db350e96c711433/src/huggingface_hub/lfs.py#L300) to guard against https://github.com/huggingface/datasets/issues/3872

**Comment by Wauplin** [comment]
We could but I'm not sure to have witness a 504 on S3 before. The issue reported in https://github.com/huggingface/datasets/issues/3872 is a 504 on the `/upload` endpoint on the Hub and this is not an endpoint that is retried on [this line](https://github.com/huggingface/huggingface_hub/blob/5eefebee2c150a2df950ab710db350e96c711433/src/huggingface_hub/lfs.py#L300).

File: src/datasets/utils/hub.py [hunk 1]
```diff
@@ -1,12 +1,47 @@
...
+                        f"{hf_api.preupload_lfs_files} timed out, retrying in {sleep_time}s... [{retry/max_retries}]"
+                    )
+                    time.sleep(sleep_time)
+                    retry += 1
+            else:
+                break
+else:
+
+    def preupload_lfs_files(hf_api: HfApi, **kwargs):
+        hf_api.preupload_lfs_files(**kwargs)
```
Comment (lhoestq) [inline src/datasets/utils/hub.py:43]: hfh>=0.20 has its own retry mechanism ?

**Review by lhoestq — APPROVED** [review]
LGTM !

**Pull request merged**